controllers/apps/transformer_cluster_load_resources.go

```diff
@@ -143,15 +143,19 @@ func (t *clusterLoadRefResourcesTransformer) checkNUpdateClusterTopology(transCt
 }
 
 func withClusterTopology(cluster *appsv1alpha1.Cluster) bool {
-	return len(cluster.Spec.ClusterDefRef) > 0 && clusterLegacyCompCnt(cluster) == 0
+	return len(cluster.Spec.ClusterDefRef) > 0 && legacyClusterCompCnt(cluster) == 0 &&
+		// to be compatible with legacy clusters: clusterDefinitionRef = xxxxx, componentDefRef = abc, componentDef = xyz
+		(len(cluster.Spec.Topology) > 0 || len(cluster.Spec.Topology) == 0 && !hasLegacyClusterCompSet(cluster))
 }
 
 func withClusterUserDefined(cluster *appsv1alpha1.Cluster) bool {
-	return len(cluster.Spec.ClusterDefRef) == 0 && len(cluster.Spec.Topology) == 0 && clusterLegacyCompCnt(cluster) == 0
+	return len(cluster.Spec.ClusterDefRef) == 0 && len(cluster.Spec.Topology) == 0 && legacyClusterCompCnt(cluster) == 0 ||
+		// to be compatible with legacy clusters: clusterDefinitionRef = xxxxx, componentDefRef = abc, componentDef = xyz
+		len(cluster.Spec.ClusterDefRef) > 0 && len(cluster.Spec.Topology) == 0 && legacyClusterCompCnt(cluster) == 0 && hasLegacyClusterCompSet(cluster)
```

Comment:
The condition is too difficult to understand. To improve code readability, it is recommended to add extra parentheses. Additionally, the compatibility scenarios can be determined in a separate function.